src/Hedgehog/Extras/Test/Tripwire.hs

```diff
@@ -0,0 +1,61 @@
+{-# LANGUAGE RankNTypes #-}
+module Hedgehog.Extras.Test.Tripwire
+  ( Tripwire
+  , makeTripwire
+  , triggerTripwire
+  , resetTripwire
+  , checkTripwire
+  , isTriggered
+  ) where
+
+import           Control.Monad.IO.Class
+import           GHC.Stack
+
+import           Control.Concurrent.MVar
+import           Control.Monad
+import           Hedgehog (MonadTest)
+import qualified Hedgehog.Extras.Test.Base as H
+import qualified Hedgehog.Internal.Property as H
+import           Prelude
+
+-- | Represents a tripwire which can be triggered only once
+newtype Tripwire = Tripwire (MVar CallStack)
+
+-- | Creates a new tripwire
+makeTripwire :: MonadIO m => m Tripwire
+makeTripwire = Tripwire <$> liftIO newEmptyMVar
+
+-- | Triggers the tripwire and registers the place of the first trigger. Idempotent.
+-- Does not do do anything besides just registering the place where this function is called.
+triggerTripwire :: HasCallStack
+                => MonadIO m
+                => Tripwire
+                -> m ()
+triggerTripwire (Tripwire mv) = withFrozenCallStack $
+  void . liftIO $ tryPutMVar mv callStack
+
+-- | Restore tripwire to initial non triggered state
+resetTripwire :: MonadIO m
+              => Tripwire
+              -> m ()
+resetTripwire (Tripwire mv) = liftIO $ void $ tryTakeMVar mv
+
+-- | Check if the tripwire is triggered. Return the first trigger location.
+isTriggered :: MonadIO m
+            => Tripwire
+            -> m (Maybe CallStack)
+isTriggered (Tripwire mv) = liftIO $ tryReadMVar mv
```

Comment:
`isTriggered` sounds like it should return `m Bool`.

Perhaps call it `callStackFromTripwire` instead?

Reply:
Fair point. I'll change its type and add `getTripSite` for that.